subprojects/core-api/src/main/java/org/gradle/api/file/ImmutableFileAccessPermissions.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.file;
+
+
+import org.gradle.api.Incubating;
+
+/**
+ * Describes file and directory access permissions. Immutable version of
+ * {@link FileAccessPermissions}.
+ *
+ * @since 8.2
+ */
+@Incubating
+public interface ImmutableFileAccessPermissions {
+
+    ImmutableFileAccessPermission getUser();
+
+    ImmutableFileAccessPermission getGroup();
+
+    ImmutableFileAccessPermission getOther();
+
+    /**
+     * Converts the permissions for the various user groups to a numeric Unix permission.
+     * See {@link FileAccessPermissions#unix(String)} for details.
+     */
+    int toUnixNumeric();
```

Comment:
:x: This should return a `Provider<Integer>` instead, so it doesn't realize lazy values too early and can be used as a task input while retaining possible dependencies.

Reply:
Agree on principle, but I'm getting terribly ugly internal usages, like: